internal/host/validator.go

```diff
@@ -1498,51 +1270,81 @@ func (v *validator) shouldValidateDNSResolution(cluster *common.Cluster) (bool,
+// This is a pre-requisite before we can call importedClusterHasManagedNetworking
+// to determine whether the cluster has managed networking or not.
+func (v *validator) canDetermineImportedClusterManagedNetworking(cluster *common.Cluster) bool {
+	for _, host := range cluster.Hosts {
+		if host.APIVipConnectivity != "" {
+			var response models.APIVipConnectivityResponse
+			if err := json.Unmarshal([]byte(host.APIVipConnectivity), &response); err != nil {
+				v.log.WithError(err).Warnf("Invalid API connectivity response")
+				continue
+			}
+
+			if !response.IsSuccess {
+				continue
+			}
+
+			return true
+		}
+	}
+
+	return false
+}
+
 func domainNameToResolve(c *validationContext, name string) string {
 	return fmt.Sprintf("%s.%s.%s", name, c.cluster.Name, c.cluster.BaseDNSDomain)
 }
 
-func (v *validator) isAPIDomainNameResolvedCorrectly(c *validationContext) ValidationStatus {
+func (v *validator) isAPIDomainNameResolvedCorrectly(c *validationContext) (ValidationStatus, string) {
+	target := "API load balancer"
 	if c.infraEnv != nil {
-		return ValidationSuccessSuppressOutput
+		return ValidationSuccessSuppressOutput, ""
 	}
 	if shouldValidate, ret := v.shouldValidateDNSResolution(c.cluster); !shouldValidate {
-		return ret
-	}
-	apiDomainName := domainNameToResolve(c, constants.APIClusterSubdomain)
-	return checkDomainNameResolution(c, apiDomainName)
-}
 
-func (v *validator) printIsAPIDomainNameResolvedCorrectly(c *validationContext, status ValidationStatus) string {
+		if c.cluster != nil {
+			apiDomainName := domainNameToResolve(c, constants.APIClusterSubdomain)
+			return ret, printIsDomainNameResolvedCorrectly(c, ret, apiDomainName, target)
+		}
+		return ret, ""
+	}
 	apiDomainName := domainNameToResolve(c, constants.APIClusterSubdomain)
-	return printIsDomainNameResolvedCorrectly(c, status, apiDomainName, "API")
+	status := checkDomainNameResolution(c, apiDomainName)
+	return status, printIsDomainNameResolvedCorrectly(c, status, apiDomainName, target)
 }
 
-func (v *validator) isAPIInternalDomainNameResolvedCorrectly(c *validationContext) ValidationStatus {
+func (v *validator) isAPIInternalDomainNameResolvedCorrectly(c *validationContext) (ValidationStatus, string) {
+	target := "internal API load balancer"
 	if c.infraEnv != nil {
-		return ValidationSuccessSuppressOutput
+		return ValidationSuccessSuppressOutput, ""
 	}
 	if shouldValidate, ret := v.shouldValidateDNSResolution(c.cluster); !shouldValidate {
-		return ret
+		if c.cluster != nil {
+			apiDomainName := domainNameToResolve(c, constants.InternalAPIClusterSubdomain)
+			return ret, printIsDomainNameResolvedCorrectly(c, ret, apiDomainName, target)
+		}
 	}
 	apiInternalDomainName := domainNameToResolve(c, constants.InternalAPIClusterSubdomain)
-	return checkDomainNameResolution(c, apiInternalDomainName)
+	status := checkDomainNameResolution(c, apiInternalDomainName)
+	return status, printIsDomainNameResolvedCorrectly(c, status, apiInternalDomainName, target)
```

Comment:
I think you could simplify this as:
```go
	shouldValidate, status := v.shouldValidateDNSResolution(c.cluster)

	apiInternalDomainName := domainNameToResolve(c, constants.InternalAPIClusterSubdomain)
        if shouldValidate {
                status = checkDomainNameResolution(c, apiInternalDomainName)
        }
	return status, printIsDomainNameResolvedCorrectly(c, status, apiInternalDomainName, target)
```

Where did that cluster conditional come from? It doesn't seem to really be needed.

Reply:
I discovered during testing that if c.cluster is nil then you will get crashes when you call printIsDomainNameResolvedCorrectly, arguably I could put the defensive code in the print function. But I had problems during tests until I introduced this.

Not really sure why it did not show up in tests before this.

Reply:
Not sure that removing this condition is a good idea

Reply:
The nil cluster name stuff, I moved this to here

`func domainNameToResolve(c *validationContext, name string) string {
	if c.cluster == nil {
		return ""
	}
	return fmt.Sprintf("%s.%s.%s", name, c.cluster.Name, c.cluster.BaseDNSDomain)
}`

the original function was

`func domainNameToResolve(c *validationContext, name string) string {
	return fmt.Sprintf("%s.%s.%s", name, c.cluster.Name, c.cluster.BaseDNSDomain)
}`

so, if c.cluster were to be nil, there would be a panic

I'm not entirely sure why the solution is now sensitive to this nil value (maybe the old formatter code handled errors differently) but I had to add this to get tests to pass.